Automattic/wp-calypso
# Checkout: Remove "variant" concept from cart manager

#### Changes proposed in this Pull Request

As part of the process of simplifying the cart manager in new checkout so that it can be more easily used outside of checkout, this PR removes the concept of product variants from the `useShoppingCartManager` hook.

Previously, there was a separate state for product variant updates, as well as a separate temporary variant storage. These can be easily replaced with the current state of the cart. See each commit for an explanation of what was done and why.

Prior to this PR there was also a `changeItemVariant` function, but this was easy to generalize into a `replaceProductInCart` function that could replace any `RequestCartItem` properties.

Tangential to the above, this PR also separates some functions that manipulate shopping cart objects into their own file separate from the TypeScript type definitions, creating a better separation of TS and JS. The new location bundles those functions (and the types) with the shopping cart manager, which will make it easier in the future to move the manager to its own package.

#### Testing instructions

- Load checkout with a plan in the cart.
- Click "Edit" on the review step.
- Verify that you see the monthly and yearly plan variants and that the correct one is selected.
- Click to change the plan variant and verify that the cart updates.
- Verify that the new variant is selected and that the new product is in the cart.
